roxie/udplib/udptrr.cpp

```diff
@@ -469,81 +448,123 @@ class CReceiveManager : implements IReceiveManager, public CInterface
-                            else
-                                newTimeout = okToSend(sender);
+                            enqueueRequest(sender, msg.flowSeq, msg.sendSeq);
                             break;
 
                         case flowType::send_completed:
-                            parent.inflight += msg.packets;
-                            if (noteDone(sender) && pendingRequests)   // This && looks wrong - noteDone returning false should mean we haven't seen the completed we wanted - so current timeout still applies. Or the one below is wrong...
-                                newTimeout = sendNextOk();
-                            else
-                                newTimeout = 5000;
+                            noteDone(sender, msg);
                             break;
 
                         case flowType::request_to_send_more:
-                            parent.inflight += msg.packets;
-                            if (noteDone(sender))
-                            {
-                                if (pendingRequests)
-                                {
-                                    enqueueRequest(sender);
-                                    newTimeout = sendNextOk();
-                                }
-                                else
-                                    newTimeout = okToSend(sender);
-                            }
+                            noteDone(sender, msg);
+                            enqueueRequest(sender, msg.flowSeq+1, msg.sendSeq);
                             break;
 
                         default:
                             DBGLOG("UdpReceiver: received unrecognized flow control message cmd=%i", msg.cmd);
                         }
-                        if (newTimeout)
-                            timeoutExpires = msTick() + newTimeout;
+                    }
+                    timeout = 5000;   // The default timeout is 5 seconds if nothing is waiting for response...
+                    if (pendingPermits)
+                    {
+                        unsigned now = msTick();
+                        for (UdpSenderEntry *finger = pendingPermits; finger != nullptr; )
+                        {
+                            if (now - finger->timeStamp >= udpRequestToSendAckTimeout)
+                            {
+                                if (udpTraceLevel || udpTraceFlow || udpTraceTimeouts)
+                                    DBGLOG("permit to send timed out after %u ms, rescheduling", udpRequestToSendAckTimeout);
+                                UdpSenderEntry *next = finger->nextSender;
+                                pendingPermits.remove(finger);
+                                pendingRequests.append(finger);
+                                finger->state = flowType::request_to_send;  // Go to the back of the queue  - MORE - lets have some code to eventually give up! Or just give up here?
```

Comment:
// Should be a way for permit from a worker that goes away to eventually be removed.  A retry count, or timeout with no communication?